- added an annotation summary window

cage2 opened the pull request:
the  command   'annotate-show-annotation-summary'  popup   a  window
  containing a  textual description  of the  annotations found  in the
  annotation file (like ~/.emacs/annotations) with the format:

      -------

  /path/of/an/annotated/file

  - annotation text
  - annotation text

  /another/path/of/an/annotated/file

  - annotation text

  ...

      -------

Each annotation text is a link  that, if activated (via a mouse click,
for  example),  opens a  buffer  showing  the  annotated file  at  the
position  of  the  annotated  text (the  underlined  one)  linked  to
annotation.

Bye!!
C.

Commit:
- added an annotation summary window

  the  command   'annotate-show-annotation-summary'  popup   a  window
  containing a  textual description  of the  annotations found  in the
  annotation file (like ~/.emacs/annotations) with the format:

      -------

  /path/of/an/annotated/file

  - annotation text
  - annotation text

  /another/path/of/an/annotated/file

  - annotation text

  ...

      -------

Each annotation text is a link  that, if activated (via a mouse click,
for  example),  opens a  buffer  showing  the  annotated file  at  the
position  of  the  annotated  text (the  underlined  one)  linked  to
annotation.

Commit:
- fixed docstring.

bastibe commented on annotate.el:
```diff
@@ -896,5 +922,48 @@ file."
+(defun annotate-show-annotation-summary ()
+  "Show a summary of all the annotations in a temp buffer"
+  (interactive)
+  (cl-labels ((ellipsize (text)
```
Why is this an inline function instead of a `defun`?

cage2 replied:
Hi!

Well, i just thought this function could not be useful outside `annotate-show-annotation-summary`, if you prefer i can just refactor it into `annotate-ellipsize` or something like that.

bastibe replied:
I think I'd prefer that, yes. I understand your intention, but factoring out the function would improve readability, I think.

cage2 replied:
As you can see i have added some more local function instead! :)
My opinion is: as they are very specific to the enclosing function the use of `cl-label` allows me to write code without "pollute" the global namespace and keeping the name reasonably short (without the `annotation-` prefix), i would say this is exactly the reason the local function form exists! But using or not local functions is a matter of personal taste so: i suggest to keep the local functions here but i left this decision to you.

Bye!
C.

bastibe replied:
I'm OK with your current solution. It does make the code somewhat cleaner.

bastibe commented on annotate.el:
```diff
@@ -896,5 +922,48 @@ file."
+                           text)))
+    (with-temp-buffer-window
+     "*annotations*" nil nil
+     (with-current-buffer "*annotations*"
```
Are both `with-temp-buffer-window` and `with-current-buffer` necessary, or would one of them suffice?

cage2 replied:
I think `with-temp-buffer-window` is needed so I tried to remove `with-current-buffer` and the summary text was inserted in the selected buffer instead of the summary window just created (i mean, the one that should hosts the summary), in my case the text was inserted straight in the source code, that was creepy! :)

bastibe replied:
OK, interesting.

cage2 replied:
well, seems that with a couple of expression more i was able to get rid of the second macro (`with-current-buffer`)!

bastibe commented on annotate.el:
```diff
@@ -730,6 +737,25 @@ file."
+file."
+  (cl-first record))
+
+(defun annotate-start-annotation-dump (annotation)
```
"start" sounds like a verb. Similarly to the previous function, I would propose something like `annotate-beginning-of-annotation`.

cage2 replied:
OK!

bastibe commented on annotate.el:
```diff
@@ -730,6 +737,25 @@ file."
+(annotate-annotations-from-dump (annotate-load-annotations))). "
+  (cl-first annotation))
+
+(defun annotate-text-annotation-dump (annotation)
```
See above: How about `annotate-text-of-annotation`?

cage2 replied:
no problem!

bastibe commented on annotate.el:
```diff
@@ -128,6 +130,11 @@ major mode is a member of this list (space separated entries)."
   :type  '(repeat symbol)
   :group 'annotate)
 
+(defcustom annotate-summary-link-max-width  64
```
Why is it 64, instead of, say 72, 80, or `fill-column`, or `visual-wrap-column` (if set)?

cage2 replied:
Because it is a power of 2! ;-)

Just kidding, but yes, it was totally arbitrary. I wonder if i could just use `window-body-width` and getting rid of this variable instead.

cage2 replied:
Seems that `window-body-width` can be used! Is this solution ok for you?

bastibe replied:
Absolutely! Thank you for investigating that!

Just out of curiosity, does the summary work correctly with crazy unicode shenanigans, like emoji or combining diacritics? It doesn't have to, if you ask me, but these kinds of things always were problematic for annotate.el.

cage2 replied:
I would say is not a mess but there are rooms for improvements! :)

bastibe reviewed (changes requested):
That's really cool! I love it!

Could you post a screenshot of what it looks like, when rendered in your Emacs?

What are your thoughts on including an entire (truncated) source line instead of just the annotated text?

I have included a few comments on how to improve the code a bit prior to merging. 

Again, thank you very much!

Commit:
- renamed function 'annotate-start-annotation-dump' to
  'annotate-beginning-of-annotation'.

Commit:
- renamed 'annotate-text-annotation-dump' to 'annotate-text-of-annotation'.

Commit:
- use 'window-body-width' to truncate the  string with an ellipse in a
  summary  window, remove  the  customizable, but  static, max  length
  value.

cage2 commented:
> That's really cool! I love it!
> 
> Could you post a screenshot of what it looks like, when rendered in your Emacs?

Sure, no problem!

> What are your thoughts on including an entire (truncated) source line instead of just the annotated text?

yes, this would be a nice improvements, any idea of the layout to accomodate this text?

Something like that, for example?

``` text

/path/of/an/annotated/file

 *  annotated text snippet 
    - annotation text
    - annotation text
 *  another annotated text snippet 
    -   annotation text

```

and so on

> Again, thank you very much!

You're welcome! I feel i am starting to learning a little more about emacs and elisp!

Bye!
C.

bastibe commented:
> > What are your thoughts on including an entire (truncated) source line instead of just the annotated text?
> 
> yes, this would be a nice improvements, any idea of the layout to accommodate this text?

Good question. Possibly something like

```
/path/to/file
* this is a line containing an {annotated text snippet} and some more...
  - annotation text
```

We could highlight the annotated text either with ascii characters, or with overlays, just like annotate.el itself does it.

Come to think of it, the coolest solution would be an `occur-edit-mode` style buffer that lists the original, editable, source lines with their true annotations. But frankly, I have no idea whatsoever how to accomplish that, or if it is even possible.

Commit:
- improved appereance of summary window
  Using outline major mode and fixed text ellipse;
- removed 'with-current-buffer' in 'annotate-show-annotation-summary'.

Commit:
- restored the kill buffer key command.

cage2 commented:
> > > What are your thoughts on including an entire (truncated) source line instead of just the annotated text?
> > 
> > 
> > yes, this would be a nice improvements, any idea of the layout to accommodate this text?
> 
> Good question. Possibly something like
> 
> ```
> /path/to/file
> * this is a line containing an {annotated text snippet} and some more...
>   - annotation text
> ```
> We could highlight the annotated text either with ascii characters, or with overlays, just like annotate.el itself does it.

I discovered (probably re-discovered) `outline-mode` and this is what i got so far:

https://www.autistici.org/interzona/img/misc/annotate-summary-example.png

I think this is close to what you asked, i have added some more empty lines here and there just according to my very personal task, but you could have different opinion about that of course! :)
 
> Come to think of it, the coolest solution would be an `occur-edit-mode` style buffer that lists the original, editable, source lines with their true annotations. But frankly, I have no idea whatsoever how to accomplish that, or if it is even possible.

This is a very ambitious goal for sure, i do not even know where to start to accomplish that, to be honest! :)

Bye!

bastibe commented on annotate.el:
```diff
@@ -896,5 +936,80 @@ file."
+                                                (annotate-text-of-annotation annotation-field)))
+                      (annotation-begin (annotate-beginning-of-annotation annotation-field))
+                      (annotation-end   (annotate-ending-of-annotation    annotation-field))
+                      (snippet-text     (build-snippet)))
```
In general, I prefer functions to have explicit arguments instead of referring to local dynamic bindings. Is there a specific reason why you implemented `build-snippet` this way?

cage2 replied:
Yes, this is confusing at best, fixed!

bastibe commented:
This looks really cool!

Why are the annotation texts list items? Do you commonly use multiple annotations of the same text? I would have expected each text to have exactly one annotation, which would not require a list.

Commit:
- changed prefix for annotation in summary window.

cage2 commented:
> This looks really cool!

Thank you! :)
 
> Why are the annotation texts list items? Do you commonly use multiple annotations of the same text? I would have expected each text to have exactly one annotation, which would not require a list.

Sorry, the list bullet was just wrong, left there only because i forgot to remove it. Now I changed it with just  spaces.

Bye!
C.

Commit:
- added args to local function.

Commit:
- fixed argument of check for file format version.

Commit:
- added a better handling of characters with width > 1.

bastibe commented:
Thank you very very much for all your work on this!

The pull request looks good to me, now! Would you like me to merge it?

cage2 commented:
On Tue, Oct 08, 2019 at 12:40:32AM -0700, Bastian Bechtold wrote:

Hi Bastibe!

> Thank you very very much for all your work on this!

My pleasure!

> The pull request looks good to me, now! Would you like me to merge it?

I you are positive: sure!

Moreover i have (almost) ready a  couple of pull requests unrelated to
this one but i think i need your help to understand if they are a good
idea or not! :)

Bye!
C.

Commit:
- removed newlines in snippet text;
- fixed wrapping of snipped text even when the latter is trunxated;
- take into account for very small  windows when text snippet is built.

cage2 commented:
Is there time for a last minute fix? ;-)

bastibe commented:
I'll merge when you tell me you're ready. There's no rush.

> Moreover i have (almost) ready a  couple of pull requests unrelated to this one but i think i need your help to understand if they are a good idea or not! :)

Just open a pull request and I'll have a look!

Thank you for putting so much work into this project!

Commit:
- making loading multiple time this mode with no effect on the buffer

  Sometimes annotate mode is applyed  to a buffer multiple time (like,
  for example,  when desktop-mode is  used), in this  case annotations
  was loaded (and  displayed) as many times as  annotate was reloaded.
  This  patch  prevent  multiple  loading  of  annotation  (note:  not
  multiple loading of annotation-mode).

cage2 commented:
On Wed, Oct 09, 2019 at 01:21:57AM -0700, Bastian Bechtold wrote:

Hi!

> I'll merge when you tell me you're ready. There's no rush.

Ok, i am fine with merging after the last commit! :)
I think we could even made this mode global now.

Bye!!
C.

cage2 commented:
On Thu, Oct 10, 2019 at 12:51:36AM -0700, Bastian Bechtold wrote:
> Merged #44 into master.

Thank you!

C.
